File: projectBlueWater/src/main/java/com/lasthopesoftware/bluewater/client/connection/waking/specs/GivenAMachine/WhenSendingAWakeSignal.java
package com.lasthopesoftware.bluewater.client.connection.waking.specs.GivenAMachine;

import com.lasthopesoftware.bluewater.client.connection.SendPackets;
import com.lasthopesoftware.bluewater.client.connection.waking.MachineAddress;
import com.lasthopesoftware.bluewater.client.connection.waking.ServerWakeSignal;
import com.lasthopesoftware.bluewater.shared.promises.extensions.specs.FuturePromise;
import com.namehillsoftware.handoff.promises.Promise;

import org.joda.time.Duration;
import org.junit.BeforeClass;
import org.junit.Test;

import java.util.concurrent.ExecutionException;

import kotlin.Unit;

import static org.assertj.core.api.Assertions.assertThat;

public class WhenSendingAWakeSignal {

	private static final byte[] expectedBytes = { -1, -1, -1, -1, -1, -1, 1, 88, -121, -6, -111, 37, 1, 88, -121, -6, -111, 37, 1, 88, -121, -6, -111, 37, 1, 88, -121, -6, -111, 37, 1, 88, -121, -6, -111, 37, 1, 88, -121, -6, -111, 37, 1, 88, -121, -6, -111, 37, 1, 88, -121, -6, -111, 37, 1, 88, -121, -6, -111, 37, 1, 88, -121, -6, -111, 37, 1, 88, -121, -6, -111, 37, 1, 88, -121, -6, -111, 37, 1, 88, -121, -6, -111, 37, 1, 88, -121, -6, -111, 37, 1, 88, -121, -6, -111, 37, 1, 88, -121, -6, -111, 37 };

	private static byte[] sentBytes;

	@BeforeClass
	public static void before() throws ExecutionException, InterruptedException {
		final SendPackets connectionProvider = (host, port, bytes) -> {
			if (port != 9) return new Promise<>(Unit.INSTANCE);
			if (!host.equals("http://my-sleeping-beauty")) return new Promise<>(Unit.INSTANCE);
			sentBytes = bytes;
			return new Promise<>(Unit.INSTANCE);
		};

		final ServerWakeSignal serverWakeSignal = new ServerWakeSignal(connectionProvider);
		new FuturePromise<>(serverWakeSignal.promiseWakeSignal(
			new MachineAddress(
				"http://my-sleeping-beauty",
				"01-58-87-FA-91-25"),
			1,
			Duration.ZERO)).get();
	}

	@Test
	public void thenTheSignalIsCorrect() {
		assertThat(sentBytes).containsExactly(expectedBytes);
	}
}

File: projectBlueWater/src/test/java/com/lasthopesoftware/bluewater/client/connection/waking/GivenAMachine/WhenSendingAWakeSignal.kt
package com.lasthopesoftware.bluewater.client.connection.waking.GivenAMachine

import com.lasthopesoftware.bluewater.client.connection.SendPackets
import com.lasthopesoftware.bluewater.client.connection.waking.MachineAddress
import com.lasthopesoftware.bluewater.client.connection.waking.ServerWakeSignal
import com.lasthopesoftware.bluewater.shared.promises.extensions.toExpiringFuture
import com.lasthopesoftware.promises.extensions.toPromise
import io.mockk.every
import io.mockk.mockk
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.BeforeAll
import org.junit.jupiter.api.Test

class WhenSendingAWakeSignal {

	private val expectedBytes = byteArrayOf(-1, -1, -1, -1, -1, -1, 1, 88, -121, -6, -111, 37, 1, 88, -121, -6, -111, 37, 1, 88, -121, -6, -111, 37, 1, 88, -121, -6, -111, 37, 1, 88, -121, -6, -111, 37, 1, 88, -121, -6, -111, 37, 1, 88, -121, -6, -111, 37, 1, 88, -121, -6, -111, 37, 1, 88, -121, -6, -111, 37, 1, 88, -121, -6, -111, 37, 1, 88, -121, -6, -111, 37, 1, 88, -121, -6, -111, 37, 1, 88, -121, -6, -111, 37, 1, 88, -121, -6, -111, 37, 1, 88, -121, -6, -111, 37, 1, 88, -121, -6, -111, 37)
	private var sentBytes: ByteArray? = null

	@BeforeAll
	fun act() {
		val sendPackets = mockk<SendPackets>()
			every { sendPackets.promiseSentPackets("http://my-sleeping-beauty", 9, any()) } answers {
				sentBytes = lastArg()
				Unit.toPromise()
			}
		ServerWakeSignal(sendPackets)
			.promiseWakeSignal(
				MachineAddress(
					"http://my-sleeping-beauty",
					"01-58-87-FA-91-25"
				)
			)
			.toExpiringFuture()
			.get()
	}

	@Test
	fun `then the signal is correct`() {
		assertThat(sentBytes).containsExactly(*expectedBytes)
	}
}
